src/Config.ts

```diff
@@ -1,30 +1,75 @@
-import { open } from "node:fs/promises";
+import { open, access, constants } from "node:fs/promises";
 import Logger from "./Logger.js";
 
-export default class Config {
-    public port: number = 25565;
+export interface Config {
+    /**
+     * Port to listen on
+     */
+    port: number;
 
     /**
      * Kick reason for when the server is shutting down
      */
-    public shutdownKickReason: string = "Server closed";
+    shutdownKickReason: string;
+}
 
+export class ConfigLoader {
     /**
      * Get a Config instance from a json file
      * @param file The file to read from
      * @returns a promise that resolves to a Config instance
      */
     public static async fromFile(file: string): Promise<Config> {
+        if (!(await ConfigLoader.exists(file))) {
+            new Logger("Config").warn("Config does not exist, creating default '%s'", file);
+            await ConfigLoader.createDefault(file);
+            return ConfigLoader.getDefault();
+        }
+
         try {
             const fd = await open(file, "r");
             const data = await fd.readFile("utf-8");
             const config = JSON.parse(data) as Config;
             fd.close();
             return config;
         } catch {
-            new Logger("Config").error("Failed to read config file, using default config");
-            return new Config();
+            new Logger("Config").error("Failed to read '%s', check your file permissions.", file);
+            process.exit(1); //TODO: better exit handling
         }
     }
```

Comment:
here we need a separate `try{}catch{}` for JSON.parse as that can fail too. We should log the error thrown by JSON.parse as that indicates what position, what character, etc.

Reply:
resolved in https://github.com/arciera/server/pull/35/commits/ff6c5f15d157b00bb2b983ca10ecd808d8c1fac3